File: app/src/main/java/com/example/ogl4jo3/accounting/setting/categorymanagement/CategoryAdapter.java
package com.example.ogl4jo3.accounting.setting.categorymanagement;

import android.app.FragmentManager;
import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.daimajia.swipe.SwipeLayout;
import com.example.ogl4jo3.accounting.R;

import java.util.List;

/**
 * 類別itemAdapter
 * <p>
 * Created by ogl4jo3 on 2017/7/14.
 */

public abstract class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.ViewHolder> {

	private Context mContext;
	private FragmentManager fragmentManager;
	private List<Category> categoryList;

	public CategoryAdapter(Context context, FragmentManager fragmentManager,
	                       List<Category> categoryList) {
		this.mContext = context;
		this.fragmentManager = fragmentManager;
		this.categoryList = categoryList;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		//return null;

		View view = LayoutInflater.from(parent.getContext())
				.inflate(R.layout.item_category, parent, false);

		return new CategoryAdapter.ViewHolder(view);
	}

	@Override
	public void onBindViewHolder(final ViewHolder holder, int position) {
		holder.tvName.setText(categoryList.get(position).getName());
		holder.ivIcon.setImageResource(categoryList.get(position).getIcon());

		holder.swipeLayoutCategory.getSurfaceView().setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				onItemClick(holder.getAdapterPosition());
			}
		});
		holder.tvDelete.setOnClickListener(new View.OnClickListener() {

			@Override
			public void onClick(View v) {
				onItemDismiss(holder.getAdapterPosition());
			}
		});
	}

	/*public void onItemMove(int fromPosition, int toPosition) {
		Collections.swap(categoryList, fromPosition, toPosition);
		notifyItemMoved(fromPosition, toPosition);
	}

	public void onItemDismiss(int position) {
		categoryList.remove(position);
		notifyItemRemoved(position);
	}*/
	public abstract void onItemClick(int position);

	public abstract void onItemMove(int fromPosition, int toPosition);

	public abstract void onItemDismiss(int position);

	@Override
	public int getItemCount() {
		return categoryList.size();
	}

	static class ViewHolder extends RecyclerView.ViewHolder {

		SwipeLayout swipeLayoutCategory;
		TextView tvName;
		ImageView ivIcon;
		TextView tvDelete;

		public ViewHolder(View itemView) {
			super(itemView);
			swipeLayoutCategory = (SwipeLayout) itemView.findViewById(R.id.swipeLayout_category);
			tvName = (TextView) itemView.findViewById(R.id.tv_name);
			ivIcon = (ImageView) itemView.findViewById(R.id.iv_icon);
			tvDelete = (TextView) itemView.findViewById(R.id.tv_delete);
			swipeLayoutCategory.setShowMode(SwipeLayout.ShowMode.PullOut);
			swipeLayoutCategory.addDrag(SwipeLayout.DragEdge.Right,
					swipeLayoutCategory.findViewById(R.id.tv_delete));
		}
	}
}

File: app/src/main/kotlin/com/ogl4jo3/accounting/ui/categoryMgmt/CategoryAdapter.kt
package com.ogl4jo3.accounting.ui.categoryMgmt

import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.daimajia.swipe.SwipeLayout
import com.google.android.material.snackbar.Snackbar
import com.ogl4jo3.accounting.R
import com.ogl4jo3.accounting.data.Category
import com.ogl4jo3.accounting.databinding.ItemCategoryBinding

class CategoryAdapter(val viewModel: CategoryMgmtViewModel) :
    ListAdapter<Category, CategoryViewHolder>(CategoryDiffCallback()) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        return CategoryViewHolder(
            ItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        ).apply {
            binding.apply {
                swipeLayoutCategory.showMode = SwipeLayout.ShowMode.PullOut
                swipeLayoutCategory.addDrag(SwipeLayout.DragEdge.Right, tvDelete)
            }
        }
    }

    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        val category = getItem(position)
        holder.binding.apply {
            this.category = category
            ivCategoryIcon.setImageResource(
                category.iconResName.drawableId(
                    this.root.context as Activity?,
                    this.root.resources
                )
            )
            swipeLayoutCategory.surfaceView.setOnClickListener {
                viewModel.navigateToItem(category)
            }
            tvDelete.setOnClickListener {
                viewModel.deleteCategory(category, {
                    Snackbar.make(
                        root,
                        root.context.getString(R.string.msg_category_deleted, category.name),
                        Snackbar.LENGTH_SHORT
                    ).show()
                }, {})
            }
            executePendingBindings()
        }
    }

    fun onItemMove(fromPosition: Int, toPosition: Int) {
        val fromItem = getItem(fromPosition)
        val toItem = getItem(toPosition)
        viewModel.swapCategoryOrderNumber(fromItem, toItem)
        //FIXME: workaround, check how to fix order number bug
        notifyItemMoved(fromPosition, toPosition)
    }
}

class CategoryViewHolder(val binding: ItemCategoryBinding) :
    RecyclerView.ViewHolder(binding.root)

class CategoryDiffCallback : DiffUtil.ItemCallback<Category>() {
    override fun areItemsTheSame(
        oldItem: Category, newItem: Category,
    ): Boolean = oldItem == newItem

    override fun areContentsTheSame(
        oldItem: Category, newItem: Category,
    ): Boolean = oldItem.id == newItem.id
}